docs: lightclient dev guide overview section

damiannolan opened the pull request:
<!-- < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < ☺
v                               ✰  Thanks for creating a PR! ✰
v    Before smashing the submit button please review the checkboxes.
v    If a checkbox is n/a - please still include it but + a little note why
☺ > > > > > > > > > > > > > > > > > > > > > > > > > > > > > > > > >  -->

## Description

Adding introduction materials for client interfaces:
- ClientState
- ConsensusState
- Height

closes: #1850 


### Commit Message / Changelog Entry

```bash
docs: adding lightclient developer guide overview
```

see the [guidelines](https://github.com/cosmos/ibc-go/blob/main/CONTRIBUTING.md#commit-messages) for commit messages. (view raw markdown for examples)


<!--
Example commit messages:

fix: skip emission of unpopulated memo field in ics20
deps: updating sdk to v0.46.4
chore: removed unused variables
e2e: adding e2e upgrade test for ibc-go/v6
docs: ics27 v6 documentation updates
feat: add semantic version utilities for e2e tests
feat(api)!: this is an api breaking feature
fix(statemachine)!: this is a statemachine breaking fix
-->

---

Before we can merge this PR, please make sure that all the following items have been
checked off. If any of the checklist items are not applicable, please leave them but
write a little note why.

- [ ] Targeted PR against correct branch (see [CONTRIBUTING.md](https://github.com/cosmos/ibc-go/blob/master/CONTRIBUTING.md#pr-targeting)).
- [ ] Linked to Github issue with discussion and accepted design OR link to spec that describes this work.
- [ ] Code follows the [module structure standards](https://github.com/cosmos/cosmos-sdk/blob/main/docs/docs/building-modules/10-structure.md).
- [ ] Wrote unit and integration [tests](https://github.com/cosmos/ibc-go/blob/master/CONTRIBUTING.md#testing).
- [ ] Updated relevant documentation (`docs/`) or specification (`x/<module>/spec/`).
- [ ] Added relevant `godoc` [comments](https://blog.golang.org/godoc-documenting-go-code).
- [ ] Provide a [commit message](https://github.com/cosmos/ibc-go/blob/main/CONTRIBUTING.md#commit-messages) to be used for the changelog entry in the PR description for review.
- [ ] Re-reviewed `Files changed` in the Github PR explorer.
- [ ] Review `Codecov Report` in the comment section below once CI passes.

damiannolan commented on docs/.vuepress/config.js:
```diff
@@ -300,7 +300,7 @@ module.exports = {
               title: "Overview",
               directory: false,
-              path: "/light-clients/overview.html",
+              path: "/ibc/light-clients/overview.html",
```
Need to fix other pages too, 404's were being returned when rendering docs with `make view-docs`

damiannolan commented:
I'll try to add something specific to lightclient creation and how it interacts via `02-client` using protobuf Any encoding of interface types

tmsdkeys reviewed (commented):
Great work to get this started @damiannolan!

Left some comments and also have a more general question about the light client development process. Namely, is there a recommended order for the light client dev to implement the interfaces and methods or not really? If there is (maybe the scaffolding was already inspired by it?), it could be useful to identify it so that the whole task seems less intimidating. Divide and conquer...

tmsdkeys commented on docs/ibc/light-clients/overview.md:
```diff
@@ -1,3 +1,45 @@
+
+# Overview 
+
+Learn how to build IBC lightclient modules and fulfill the interfaces required to integrate with core IBC. {synopsis}
```
light client instead of lightclient?

tmsdkeys commented on docs/ibc/light-clients/overview.md:
```diff
@@ -1,3 +1,45 @@
+# Overview 
+
+Learn how to build IBC lightclient modules and fulfill the interfaces required to integrate with core IBC. {synopsis}
+
```
Even though later on we assume that the reader has prior knowledge, I think it's best practice to concisely summarize what a light client does and what the context is it operates in (trust-minimized interoperability so we need light clients to track the consensus of a counterparty chain)

tmsdkeys commented on docs/ibc/light-clients/overview.md:
```diff
@@ -1,3 +1,45 @@
+- Constraints used for misbehaviour detection.
+- Constraints used for state verification.
+- Constraints used for client upgrades.
+
```
Does it make sense here to mention that some properties are customizable, and others are fixed during the client lifecycle?

damiannolan replied:
After taking a browse through some of the other existing docs I've noticed fields for tendermint client upgrades are documented [here](https://ibc.cosmos.network/main/ibc/upgrades/quick-guide.html#ibc-client-breaking-upgrades). I think we can focus in on mutability of client state fields in the dedicated client upgrades page (upgrades.md) and link to the existing upgrade docs for tm clients there also

tmsdkeys commented:
Didn't spot the comment you left earlier @damiannolan 
> Happy to add more sections etc, for example if people think we need to define what misbehaviour is within the overview section or if its okay to leave it to the dedicated misbehaviour page.

In my mind, an overview page should contain a coherent and full high-level story, satisfying two potential needs:

1. A general introduction for a developer who intends to start development work and is expected to check out the dedicated pages, reflecting the outline of the dedicated sections and putting them in its larger context
2. A brief overview for the reader who's interested in knowing what light client development entails, yet will not go into actual development (could be a product person or engineering manager trying to gage what development efforts are required). they may not want to visit dedicated pages

So I would include another section to outline the most important use cases of the light client, which afaik (correct me if I'm wrong) are: updating consensus state and submitting misbehaviour through `ClientMessage` and state proof verification. I would mention these and refer to the respective sections for more details

colin-axner commented on docs/ibc/light-clients/overview.md:
```diff
@@ -1,3 +1,45 @@
+
+Learn how to build IBC lightclient modules and fulfill the interfaces required to integrate with core IBC. {synopsis}
+
+The following aims to provide a high level IBC lightclient module developer guide. The lifecycle of IBC lightclients are created and maintained by the core IBC `MsgServer` and the abstractions provided by the `02-client` submodule. A lightclient module developer is only required to implement the methods and fulfil the expectations of a defined set of interfaces. In ibc-go these interfaces reside  ...
```
I'm a bit confused by the terminology "The lifecycle of IBC light clients". To me, "lifecycle" implies a beginning and end (packets have a clear beginning and end), but light clients aren't meant to have an end. Is there different wording we could use?

damiannolan replied:
Yeah good point. I'll use your suggestion in the other comment! 👍

colin-axner commented on docs/ibc/light-clients/overview.md:
```diff
@@ -1,3 +1,45 @@
+
+Learn how to build IBC lightclient modules and fulfill the interfaces required to integrate with core IBC. {synopsis}
+
+The following aims to provide a high level IBC lightclient module developer guide. The lifecycle of IBC lightclients are created and maintained by the core IBC `MsgServer` and the abstractions provided by the `02-client` submodule. A lightclient module developer is only required to implement the methods and fulfil the expectations of a defined set of interfaces. In ibc-go these interfaces reside  ...
```
Maybe something like this instead?
```suggestion
The following aims to provide a high level IBC light client module developer guide. Access to IBC light clients are gated by the core IBC `MsgServer` which utilizes the abstractions set by the `02-client` submodule to call into a light client module. A light client module developer is only required to implement the `ClientState` and `ConsensusState` interfaces as defined in the `core/modules/exported` package. 
```
Should probably also link to the interfaces

damiannolan replied:
Thanks, I will use this and adapt it slightly. I guess this part is actually incorrect:
> A light client module developer is only required to implement the `ClientState` and `ConsensusState` interfaces as defined in the `core/modules/exported` package. 

`ClientMessage` also needs to be implemented for updates and misbehaviour

colin-axner commented on docs/ibc/light-clients/overview.md:
```diff
@@ -1,3 +1,45 @@
+
+## Concepts and vocabulary
+
+This guide serves to be implementation specific with respect to ibc-go. Readers are expected to familiarize themselves with the IBC protocol specifications.
```
This guide is meant to be specific to ibc-go. It is not an intro to IBC light clients. It is a "how to develop light clients with ibc-go". I agree asking the reader to familiarize themselves with the spec isn't particularly useful, I'd rather link to an IBC overview explainer 

In general, ibc-go should only host documentation specific to interacting with ibc-go. The reasoning here is that ibc-go will not be the only IBC implementation, while it is at the moment, it'd be redundant if every IBC implementation hosted documentation explaining IBC concepts. 

There should be a place with:
- general easy to understand IBC concepts
- specs (cosmos/ibc)
- implementation specific documentation (ibc-go for example)

colin-axner commented on docs/ibc/light-clients/overview.md:
```diff
@@ -1,3 +1,45 @@
+
+### ClientState 
+
+ClientState is a term used to define the data structure which encapsulates opaque lightclient state. This refers to internal data that accommodate the semantics of a lightclient. This may be any arbitrary data such as:
```
> This refers to internal data that accommodate the semantics of a lightclient.

I find this phrasing a bit broad. What do you mean by "internal data that accommodate the semantics"?

damiannolan replied:
Yeah, I made this intentionally broad as it seemed pretty broad from the spec. But what I was referring to is the fields on the type which make up the constraints/rules listed below:

```
- Constraints used for client updates.
- Constraints used for misbehaviour detection.
- Constraints used for state verification.
- Constraints used for client upgrades.
```

Happy to change or add more info if you have any suggestions. I can try to think of how to rephrase it also.

damiannolan replied:
Awesome, thanks for the suggestion. I've gone with the following for now, we can always revisit this when we get some user feedback.

> ClientState is a term used to define the data structure which encapsulates opaque light client state. The `ClientState` contains all the information needed to verify a `ClientMessage` and perform membership and non-membership proof verification of counterparty state. This includes properties that refer to the remote state machine, the light client type and the specific light client instance.

colin-axner commented on docs/ibc/light-clients/overview.md:
```diff
@@ -1,3 +1,45 @@
+- Constraints used for state verification.
+- Constraints used for client upgrades.
+
+The `ClientState` type maintained within the lightclient module *must* implement the `ClientState` interface defined in `core/modules/exported/client.go`.
```
we should add links to the client interface

colin-axner reviewed (commented):
Nice start! Should we add a section which explains how 02-client interacts with the `ClientState` and `ConsensusState`?

damiannolan commented:
Thanks for the feedback @tmsdkeys and @colin-axner. I'm going to convert this back to draft and I'll begin addressing comments and continuing work on this

colin-axner commented on docs/ibc/light-clients/overview.md:
```diff
@@ -1,3 +1,68 @@
+
+A light client module developer should be concerned with three main interfaces:
+
+- [`ClientState`](https://github.com/cosmos/ibc-go/blob/main/modules/core/exported/client.go#L36) encapsulates the light client implementation and its semantics.
+- [`ConsensusState`](https://github.com/cosmos/ibc-go/blob/main/modules/core/exported/client.go#L134) tracks consensus data used for verification of client updates, misbehaviour detection and proof verification of counterparty state.
+- [`ClientMessage`](https://github.com/cosmos/ibc-go/blob/main/modules/core/exported/client.go#L148) used for submitting block headers for client updates and submission of misbehaviour evidence using conflicting headers. 
```
The links should link to our docs in the light clients guide, also would avoid links to main

colin-axner reviewed (approved):
LGTM! Very nice work. The light client docs are coming together!
